File: AOC2021.cs/Day2.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AOC2021.cs
{
    internal class Day2
    {
        public static int Resolve(int _part, string _filename)
        {
            KeyValuePair<string, int>[] data;
            try
            {
                data = File.ReadAllText(_filename)
                    .Split('\n')
                    .Select(s =>
                    {
                        int n;
                        string[] parts = s.Split(' ');
                        return new KeyValuePair<string, int>(
                            parts[0], 
                            int.TryParse(
                                parts.Count() > 1 
                                ? parts[1]
                                : "0", out n) 
                            ? n 
                            : 0);
                    })
                    .ToArray();
            }
            catch (Exception)
            {
                return -1;
            }

            switch (_part)
            {
                case 1:
                    return Day2_1(data);
                case 2:
                    return Day2_2(data);
                default:
                    return -1;
            }
        }

        static int Day2_1(KeyValuePair<string, int>[] _dayData)
        {
            int depth = 0;
            int horizontalPosition = 0;
            foreach (KeyValuePair<string, int> kvp in _dayData)
            {
                switch (kvp.Key)
                {
                    case "forward":
                        horizontalPosition += kvp.Value;
                        break;
                    case "up":
                        depth -= kvp.Value;
                        break;
                    case "down":
                        depth += kvp.Value;
                        break;
                    default:
                        break;
                }
            }
            return depth * horizontalPosition;
        }

        static int Day2_2(KeyValuePair<string, int>[] _dayData)
        {
            int depth = 0;
            int horizontalPosition = 0;
            int aim = 0;
            foreach (KeyValuePair<string, int> kvp in _dayData)
            {
                switch (kvp.Key)
                {
                    case "forward":
                        horizontalPosition += kvp.Value;
                        depth += aim * kvp.Value;
                        break;
                    case "up":
                        aim -= kvp.Value;
                        break;
                    case "down":
                        aim += kvp.Value;
                        break;
                    default:
                        break;
                }
            }
            return depth * horizontalPosition;
        }
    }
}

File: AOC2021.cpp/Day2.cpp
#include "Day2.h"
#include <map>
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <iostream>

struct kvpair
{
	std::string key;
	int value;
};

int day2_1(std::vector<struct kvpair> _data);
int day2_2(std::vector<struct kvpair> _data);

int Day2::Resolve(int _part, std::string _filename)
{
	std::vector<struct kvpair> data;
	{
		std::ifstream file(_filename);
		std::stringstream stringstreamBuffer;
		std::string stringBuffer;
		size_t pos;

		stringstreamBuffer << file.rdbuf();
		stringBuffer = stringstreamBuffer.str();

		while ((pos = stringBuffer.find('\n')) != std::string::npos)
		{
			data.push_back({
				stringBuffer.substr(0, stringBuffer.find(' ')),
				std::stoi(stringBuffer.substr(stringBuffer.find(' '), pos))
				});
			stringBuffer.erase(0, pos + 1);
		}
	}

	switch (_part)
	{
	case 1:
		return day2_1(data);
	case 2:
		return day2_2(data);
	default:
		return -1;
	}
}

int day2_1(std::vector<struct kvpair> _data)
{
	int depth = 0;
	int horizontalPosition = 0;
	for (size_t i = 0; i < _data.size(); i++)
	{
		if (_data[i].key._Equal("forward"))
		{
			horizontalPosition += _data[i].value;
		}
		else if (_data[i].key._Equal("up"))
		{
			depth -= _data[i].value;
		}
		else if (_data[i].key._Equal("down"))
		{
			depth += _data[i].value;
		}
	}
	return depth * horizontalPosition;
}

int day2_2(std::vector<struct kvpair> _data)
{
	int depth = 0;
	int horizontalPosition = 0;
	int aim = 0;
	for (size_t i = 0; i < _data.size(); i++)
	{
		if (_data[i].key._Equal("forward"))
		{
			horizontalPosition += _data[i].value;
			depth += aim * _data[i].value;
		}
		else if (_data[i].key._Equal("up"))
		{
			aim -= _data[i].value;
		}
		else if (_data[i].key._Equal("down"))
		{
			aim += _data[i].value;
		}
	}
	return depth * horizontalPosition;
}
